### PR TITLE
Fix `Variable.addgrad` dtype mismatch error check

### LINKED ISSUE #7136: Variable.addgrad with mismatched dtype might not cause error
Version: current mater 8565bad9334f6cc32cd70a9d79fd42f74eeec3c2
```py
import chainer
import numpy as xp

a = chainer.Variable(xp.array([1, 1], 'float32'))
#a.grad = xp.array([3, 5], 'float32')

b = chainer.Variable(xp.array([2, 2], 'float64'))
b.grad = xp.array([4, 7], 'float64')

a.addgrad(b)

print(a.dtype, a.grad.dtype)
```

An error is expected but it results in the following output (indicating mismatched gradient dtype)
```
float32 float64
```

If the comment-out is removed, it becomes an error:
```
chainer.utils.type_check.InvalidType: 
Invalid operation is performed in: _ + _ (Forward)

Expect: lhs.dtype == rhs.dtype
Actual: float64 != float32
```

### PR DESCRIPTION
**Pull request by niboshi**
Fixes #7136

### CODE REVIEW

**Comment by hvy** [comment]
The first part of the issue was resolved in https://github.com/chainer/chainer/pull/7132. Is my understanding correct that are you addressing the second part, i.e. when `a.grad` is set in this PR?

**Comment by niboshi** [comment]
What do you mean?
What are the first part and the second part?

**Comment by hvy** [comment]
First part 
`#a.grad = xp.array([3, 5], 'float32')`
Second part
`a.grad = xp.array([3, 5], 'float32')`

**Comment by hvy** [comment]
Could you also resolve conflicts?

**Comment by niboshi** [comment]
I don't understand yet. The only issue is that dtype mismatch is not reported, and no more.

**Comment by hvy** [comment]
Sorry for pinging again but could you resolve the conflicts?

**Comment by niboshi** [comment]
Sorry for being late.
Conflict resolved.

**Comment by niboshi** [comment]
TODO after this PR:
https://github.com/chainer/chainer/pull/7146/files#diff-05835e3a5c0abcd4ac2d0fb29936ac76R545

File: tests/chainer_tests/test_variable.py [hunk 1]
```diff
@@ -904,52 +904,122 @@ def check_addgrad(
...
 
 
 @testing.backend.inject_backend_tests(None, _backend_params)
-@testing.backend.inject_backend_tests(None, _nonchainerx_backend_params)
-@testing.backend.inject_backend_tests(None, _nonchainerx_backend_params)
+@testing.backend.inject_backend_tests(None, _backend_params)
+@testing.backend.inject_backend_tests(None, _backend_params)
 @addgrad_test_parameterize
-class TestVariableAddgradNonChainerx(
-        VariableAddgradTestBase, unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/test_variable.py]: Oh, thank you for pointing out!
Fixed.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Can you check the CI failures? Seems to fail when trying to access the `grad` attributes.

**Pull request closed without merging**